s2argv.3
.\"* s2argv: convert strings to argv
.\" Copyright (C) 2014 Renzo Davoli. University of Bologna. <renzo@cs.unibo.it>
.\" 
.\" This library is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU Lesser General Public
.\" License as published by the Free Software Foundation; either
.\" version 2.1 of the License, or (at your option) any later version.
.\" 
.\" This library is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Lesser General Public License for more details.
.\" 
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with this library; if not, write to the Free Software
.\" Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
.TH S2ARGV 3 2014-05-27 "VirtualSquare" "Linux Programmer's Manual"
.SH NAME
s2argv \- convert a command string in an argv array
.SH SYNOPSIS
.B #include <unistd.h>
.br
.B #include <stdlib.h>
.br
.B #include <execs.h>
.sp
.BI "char **s2argv(const char *" args ");"
.br
.BI "void s2argv_free(char **" argv ");"
.sp
.br
.BI "size_t s2argvlen(char **" argv ");"
.sp
.br
.BI "size_t s2argc(char **" argv ");"
.sp
.br
.BI "typedef char * (* s2argv_getvar_t) (const char *name);"
.br
.BI "extern s2argv_getvar_t s2argv_getvar;"
.sp
These functions are provided by libexecs and libeexecs. Link with \fI-lexecs\fR or \fI-leexecs\fR.
.sp
.SH DESCRIPTION
.BR s2argv
convert a command string in an argv array for \fBexecv\fR(3), 
\fBexecvp\fR(3) or \fBexecvpe\fR(3).
Single or double quotes can be used to delimitate command arguments including
spaces and a non quoted backslash (\fB\e\fP)
is the escape character to protect the next char.
.br
\fBs2argv\fR can parse several commands separated by semicolons (\fB;\fR).
The argv of each command is terminated by a NULL element, one further NULL element
tags the end of the array returned by s2argv.
.br
\fBs2argv\fR supports variables as arguments. When an argument of a command is a dollar
sign followed by a name (e.g. $USER) \fBs2argv\fR puts the output of the \fBs2argv_getvar\fR
function instead. (The default value is \fBs2argv_getvar=getenv\fR.
Programmers can use their own custom function instead).
.sp
.BR s2argv_free
frees the memory that was allocated by \fBs2argv\fR.
.sp
.BR s2argvlen
returns the length of the array returned by \fBs2argv\fR.
.sp 
.BR s2argc
returns the number of arguemnts of the (first) command returned by \fBs2argv\fR.
(The beginning of the next argv is \fBargv+s2argc(argv)+1\fR).
.SH RETURN VALUE
.BR s2argv
returns a dynamically allocated argv, ready to be used as an argument to
\fBexecv\fR(3), \fBexecvp\fR(3) or \fBexecvpe\fR(3).
The return value of
.BR s2argv
should be freed by
.BR s2argv_free
in case the exec command does not succeed.
.SH EXAMPLE
The following program demonstrates the use of \fBs2argv\fR:
.BR
.sp
\&
.nf
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <execs.h>

#define BUFLEN 1024
int main(int argc, char *argv)
{
	char buf[BUFLEN];
	printf("type in a command and its arguments, e.g. 'ls -l'\\n");
	if (fgets(buf, BUFLEN, stdin) != NULL) {
		char **argv=s2argv(buf);
		execvp(argv[0], argv);
		s2argv_free(argv);
		printf("exec error\\n");
	}
}
.fi
.SH SEE ALSO
.BR exec (3)
.SH BUGS
Bug reports should be addressed to <info@virtualsquare.org>
.SH AUTHOR
Renzo Davoli <renzo@cs.unibo.it>



s2argv.c
/*
 * s2argv: convert strings to argv
 * Copyright (C) 2014 Renzo Davoli. University of Bologna. <renzo@cs.unibo.it>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */
#define _GNU_SOURCE
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <s2argv.h>
#include <unistd.h>
//#define  S2ARGVTEST

//char *sn[] = { "END", "SPACE", "CHAR", "SGLQ", "DBLQ", "ESCAPE", "SEMIC", "VAR", "ESCVAR", "DBLESC" };
#define END 0
#define SPACE 1
#define CHAR 2
#define SGLQ 3
#define DBLQ 4
#define ESCAPE 5
#define SEMIC 6
#define VAR 7
#define ESCVAR 8
#define DBLESC 9
#define NSTATES (DBLESC+1)

#define NEWARG 0x1
#define CHCOPY 0x2
#define ENDARG 0x4
#define ENDVAR 0x8
#define ENDCMD 0x10

char nextstate[NSTATES][NSTATES-1]= {
	{END,    0,   0,   0,   0,     0,    0,   0}, // END
	{END,SPACE,CHAR,SGLQ,DBLQ,ESCAPE,SEMIC, VAR}, // SPACE
	{END,SPACE,CHAR,SGLQ,DBLQ,ESCAPE,SEMIC,CHAR}, // CHAR
	{END, SGLQ,SGLQ,CHAR,SGLQ,  SGLQ, SGLQ,SGLQ}, // SGLQ
	{END, DBLQ,DBLQ,DBLQ,CHAR,DBLESC, DBLQ,DBLQ}, // DBLQ
	{END, CHAR,CHAR,CHAR,CHAR,  CHAR, CHAR,CHAR}, // ESCAPE
	{END,SEMIC,CHAR,SGLQ,DBLQ,ESCAPE,SEMIC, VAR}, // SEMIC
	{END,SPACE, VAR, VAR, VAR,   VAR,SEMIC, VAR}, // VAR
	{END,  VAR, VAR, VAR, VAR,   VAR,  VAR, VAR}, // ESCVAR
	{END, DBLQ,DBLQ,DBLQ,DBLQ,  DBLQ, DBLQ,DBLQ}}; // DBLESC

char action[NSTATES][NSTATES-1]= {
	{ENDCMD|     0,     0,            0,     0,     0,     0,            0,     0}, //END
	{ENDCMD|     0,     0,NEWARG|CHCOPY,NEWARG,NEWARG,NEWARG,       ENDCMD,NEWARG}, //SPACE
	{ENDCMD|ENDARG,ENDARG,       CHCOPY,     0,     0,     0,ENDCMD|ENDARG,     0}, //CHAR
	{ENDCMD|ENDARG,CHCOPY,       CHCOPY,     0,CHCOPY,CHCOPY,       CHCOPY,CHCOPY}, //SNGQ
	{ENDCMD|ENDARG,CHCOPY,       CHCOPY,CHCOPY,     0,CHCOPY,       CHCOPY,CHCOPY}, //DBLQ
	{ENDCMD|ENDARG,CHCOPY,       CHCOPY,CHCOPY,CHCOPY,CHCOPY,       CHCOPY,CHCOPY}, //ESCAPE
	{ENDCMD|     0,     0,NEWARG|CHCOPY,NEWARG,NEWARG,NEWARG,            0,NEWARG}, //SEMIC
	{ENDCMD|ENDVAR,ENDVAR,       CHCOPY,     0,     0,     0,ENDCMD|ENDVAR,     0}, //VAR
	{ENDCMD|ENDVAR,CHCOPY,       CHCOPY,CHCOPY,CHCOPY,CHCOPY,       CHCOPY,CHCOPY}, //ESCVAR
	{ENDCMD|ENDARG,CHCOPY,       CHCOPY,CHCOPY,CHCOPY,CHCOPY,       CHCOPY,CHCOPY}}; //DBLESC

s2argv_getvar_t s2argv_getvar=getenv;
int (* s2_fork_security)(void *s2_fork_security_arg);
void *s2_fork_security_arg;

static int args_fsa(const char *args, char **argv, char *buf)
{
	int state=SPACE;
	int argc=0;
	char *thisarg=NULL;
	for (;state != END;args++) {
		int this;
		switch (*args) {
			case 0:
				this=END;
				break;
			case ' ':
			case '\t':
			case '\n':
				this=SPACE;
				break;
			case '\'':
				this=SGLQ;
				break;
			case '"':
				this=DBLQ;
				break;
			case '\\':
				this=ESCAPE;
				break;
			case ';':
				this=SEMIC;
				break;
			case '$':
				this=VAR;
				break;
			default:
				this=CHAR;
		}
		if (argv) {
			if (action[state][this] & NEWARG)
				thisarg=buf;
			if (action[state][this] & CHCOPY)
				*buf++=*args;
			if (action[state][this] & ENDARG) {
				*buf++=0;
				*argv++=thisarg;
			}
			if (action[state][this] & ENDVAR) {
				*buf++=0;
				if (s2argv_getvar) {
					*argv=s2argv_getvar(thisarg);
					if (*argv == NULL)
						*argv="";
				} else
					*argv="";
				argv++;
			}
			if (action[state][this] & ENDCMD) 
				*argv++=0;
		} 
		if (action[state][this] & (ENDARG|ENDVAR))
			argc++;
		if (action[state][this] & ENDCMD)
			argc++;
		//printf("%c %s+%s=%s %x\n",*args,sn[state],sn[this],sn[nextstate[state][this]],action[state][this]);
		//printf("%s %d->%d\n",args,state,nextstate[state][this]);
		state=nextstate[state][this];
	}
	if (argv)
		*argv=0;
	return argc;
}

#ifndef NOCOPY_ONLY

char **s2argv(const char *args)
{
	int argc=args_fsa(args,NULL,NULL);
	char buf[strlen(args)+1];
	char **argv=calloc(argc+1,sizeof(char *));
	if (argv) {
		int i;
		args_fsa(args,argv,buf);
		for (i=0; i<argc+1; i++)
			argv[i]=argv[i]?strdup(argv[i]):0;
		for (i=0; i<argc+1; i++)
			printf("%d %s\n",i,argv[i]);
	}
	return argv;
}

void s2argv_free(char **argv)
{
	char **scan;
	for (scan=argv; *scan != (char *)0; scan++) {
		for (; *scan != (char *)0; scan++)
			free(*scan);
	}
	free(argv);
}

size_t s2argvlen(char **argv) {
	int len;
	for (len=0; *argv; len++, argv++)
		for(; *argv; len++, argv++)
			;
	return len;
}

size_t s2argc(char **argv) {
	int argc;
	for (argc=0; *argv; argc++, argv++)
		;
	return argc;
}

int s2multiargv(const char *args, int (*f)(char **argv, void *opaque), void *opaque)
{
	int argc=args_fsa(args,NULL,NULL);
	char *argv[argc+1];
	char buf[strlen(args)+1];
	char **thisargv=argv;
	args_fsa(args,argv,buf);
	int rv=0;
	while (*thisargv && rv==0) {
		rv=f(thisargv, opaque);
		while (*thisargv) thisargv++;
		thisargv++;
	}
	return rv;
}
#endif

int execs_common(const char *path, const char *args, char *const envp[], char *buf)
{
	int argc=args_fsa(args,NULL,NULL);
	char *argv[argc+1];
	args_fsa(args,argv,buf);
	if (path)
		return execve(path, argv, envp);
	else
		return execvpe(argv[0], argv, envp);
}

#ifdef S2ARGVTEST

static void printargv(char **argv)
{
	for(;*argv!=0;argv++) {
		int argc=0;
		for(;*argv!=0;argv++,argc++)
			printf("argv[%d]=\"%s\"\n",argc,*argv);
		argc++;
	}
}

static int print1argv(void *useless, char **argv) 
{
	int argc=0;
	for(;*argv!=0;argv++,argc++)
		printf("argv[%d]=\"%s\"\n",argc,*argv);
	return 0;
}

int main()
{
	char buf[1024];
	s2argv_getvar=getenv;
	while (1) {
		char **myargv;
		fgets(buf,1024,stdin);
		//buf[strlen(buf)-1]=0;
#if 0
		myargv=s2argv(buf);
		printargv(myargv);
		printf("len %ld argc %ld\n", s2argvlen(myargv), s2argc(myargv));
		s2argv_free(myargv);
		s2multiargv(NULL, buf, print1argv);
		if (fork()==0) {
			execsp_nocopy(buf);
			exit(-1);
		} else {
			int status;
			wait(&status);
		}
#else
		printf("%d\n",WEXITSTATUS(system_execsa(buf)));
#endif
	}
}
#endif
